crates/astria-sequencer/src/app/mod.rs

```diff
@@ -148,35 +149,171 @@ const POST_TRANSACTION_EXECUTION_RESULT_KEY: &str = "post_transaction_execution_
+            }
+            let partial_proposal = abci::request::ProcessProposal {
+                hash: Hash::default(),
+                ..proposal.clone()
+            };
+            let data = pb::RequestProcessProposal::from(partial_proposal).encode_to_vec();
+
+            self.sha = Some(Sha256::digest(data).into());
+        }
+
+        if self.hash.is_some() {
+            bail!("full proposal already set");
+        }
+
+        self.hash = Some(proposal.hash);
+        Ok(())
+    }
+
+    // If the proposal has been set return if this is valid or not
+    fn validate_proposal(
+        &self,
+        request: &abci::request::ProcessProposal,
+    ) -> ExecutedProposalValidationStatus {
+        use sha2::Sha256;
+        use tendermint_proto::v0_38::abci as pb;
+
+        if self.sha.is_none() {
+            return ExecutedProposalValidationStatus::Unset;
+        }
+
+        let proposal = abci::request::ProcessProposal {
+            hash: Hash::default(),
+            proposed_last_commit: request.proposed_last_commit.clone(),
+            height: request.height,
+            time: request.time,
+            proposer_address: request.proposer_address,
+            next_validators_hash: request.next_validators_hash,
+            misbehavior: request.misbehavior.clone(),
+            txs: request.txs.clone(),
+        };
+        let data = pb::RequestProcessProposal::from(proposal).encode_to_vec();
+        let sha: [u8; 32] = Sha256::digest(data).into();
+
+        if self.sha == Some(sha) {
+            ExecutedProposalValidationStatus::Valid
+        } else {
+            ExecutedProposalValidationStatus::Invalid
+        }
+    }
+
+    fn validate_finalization(
+        &self,
+        finalize_block: &FinalizeBlock,
+    ) -> ExecutedProposalValidationStatus {
+        use sha2::Sha256;
+        use tendermint_proto::v0_38::abci as pb;
+
+        if self.hash.is_none() || self.sha.is_none() {
+            return ExecutedProposalValidationStatus::Unset;
+        }
+        if self.hash != Some(finalize_block.hash) {
+            return ExecutedProposalValidationStatus::Invalid;
+        }
+        let proposal = abci::request::ProcessProposal {
+            hash: Hash::default(),
+            proposed_last_commit: Some(finalize_block.decided_last_commit.clone()),
+            height: finalize_block.height,
+            time: finalize_block.time,
+            proposer_address: finalize_block.proposer_address,
+            next_validators_hash: finalize_block.next_validators_hash,
+            misbehavior: finalize_block.misbehavior.clone(),
+            txs: finalize_block.txs.clone(),
+        };
+        let data = pb::RequestProcessProposal::from(proposal).encode_to_vec();
+        let sha: [u8; 32] = Sha256::digest(data).into();
+
+        if self.sha == Some(sha) {
+            ExecutedProposalValidationStatus::Valid
+        } else {
+            ExecutedProposalValidationStatus::Invalid
         }
```

Comment:
don't think this is necessary - in finalize_block, checking the block hashes match is sufficient, if any of the other fields don't match then the hash won't